### DISCUSSION TITLE
Need some help...

### DISCUSSION THREAD

**Discussion by Akascape** [nmbr73/Shadertoys#47] [Q&A] [answered]
Hi @nmbr73 @J-i-P-i , I am big fan of your shader fuses. Inspired from those, I decided to make my own fuse plugin. I really learned a lot from [this tutorial](https://www.youtube.com/watch?v=dbrPWRldmbs) @nmbr73 made, (don't know german, watched with english captions).
Then I tried to convert a small shader similar to this [legofied.fuse](https://github.com/nmbr73/Shaderfuse/blob/main/Shaders/Misc/Legofied.fuse). I was successful in converting it, also added custom sliders and options. It was perfectly rendering the shader and everything was working fine on my machine.
I am using Davinci Resolve 18 with AMD graphics. (processing mode: OpenCL)

But then I shared it to a different machine where it has a cuda graphics card. Unfortunately, the plugin was not working there even after changing the graphics mode to OpenCL.  The error message shown in the console was this: 
```
RT kernel DCTLKernel failed to build:
DCTLKernel(3151): error: no suitable constructor exists to convert from "float" to "float2"
```
_This message is not showing on my machine having no cuda._
I looked here: https://github.com/nmbr73/Shaderfuse/issues/15 but there is no information. I don't know what to do with the compatibility issue. Please tell what I can do? I am a beginner actually :(

**Comment by J-i-P-i** [comment]
Hi Akascape
nice to hear reading your feedback.
When it comes to conversion, OpenCL is indeed the problematic option. I also found that OpenCL doesn't work the same on all machines, there are obviously different flavors. In order to be able to evaluate this error, I would need the shader file. 
Very often this error occurs when initializing a variable - while in GLSL the following line works:
vec2 Var1 = vec2(0.0f);
cannot convert this to float2 Var1 = float2(0.0f);
to be translated. but in

float2 Var1 = to_float2_s(0.0f) or float2 Var1 = to_float2(0.0f,0.0f);

Best regards

**Comment by Akascape** [comment]
@J-i-P-i Thanks for the response. 
> In order to be able to evaluate this error, I would need the shader file.

I am actually making a glitch effect plugin. Here is the full fuse file: [`SuperGlitch.fuse`](https://github.com/Akascape/Akascape/releases/download/Test/SuperGlitch.fuse)
This is the DCTL Kernel:
```C
-- Description of kernel parameters
 
DCTLParams =
[[
  float parameter1;
  float parameter2;
  float parameter3;
  float parameter4;
  float parameter5;
  float parameter6;
  float x_axis;
  float y_axis;
  int width;
  int height;
  float itime;
  int compOrder;
]]

-- source of kernel
DCTLKernel =
    [[
	

__DEVICE__ float2 quantize(float2 v, float steps) {return _floor(v * steps) / steps;}

__DEVICE__ float4 posterize(float4 color, float numColors) {return _floor(color * numColors - 0.5f) / numColors;}

//############################################################################################################################
__KERNEL__ void DCTLKernel(__CONSTANTREF__ DCTLParams *params, __TEXTURE2D__ iChannel0, __TEXTURE2D_WRITE__ dst)
{
  DEFINE_KERNEL_ITERATORS_XY(x, y);
  //---------------------------------------
  if (x < params->width && y < params->height) 
  {
  
     //-----------------------
     float2 fragCoord = to_float2(x, y); 
     float2 iResolution = to_float2(params->width, params->height);
     float itime = params->itime;
	 float4 fragColor = to_float4_s(0.0f);
     //-----------------------
    
    float GlitchAmount = params->parameter1;
   
    float2 uv = fragCoord / iResolution;
 
    float amount = _powf(GlitchAmount, 2.0f);
    float2 pixel = params->parameter6 / iResolution;    
    float4 color = _tex2DVecN(iChannel0, uv.x, uv.y, 15);
    float t = _fmod(_fmod(itime, amount * 100.0f * (amount - 0.5f)) * 100.0f, params->parameter3);
    float4 postColor = posterize(color, 16.0f);
	
    float2 i = quantize(uv, 64.0f * t) + pixel * (postColor.x - to_float2_s(0.5f)) * 100.0f;
    float2 j = quantize(uv, 32.0f - t) + pixel * (postColor.x - to_float2_s(0.5f)) * 1000.0f;
    float2 k = quantize(uv, 16.0f + t) + pixel * (postColor.x - to_float2_s(0.5f)) * 20.0f;
	
    float4 a = posterize(_tex2DVecN(iChannel0, i.x, i.y, 15), 5.0f);
    float4 b = posterize(_tex2DVecN(iChannel0, j.x, j.y, 15), 4.0f);
    float4 c = posterize(_tex2DVecN(iChannel0, k.x, k.y, 15), 16.0f);
	
    float2 p = uv.x + amount * (quantize((a * t - b + c - (t + t / 2.0f) / 10.0f).x, 16.0f) - to_float2_s(0.5f)) * pixel * params->parameter4;
    float2 q = uv.y + amount * (quantize((a * t - b + c - (t + t / 2.0f) / 10.0f).x, 16.0f) - to_float2_s(0.5f)) * pixel * params->parameter4;
	
    if (params->x_axis == 1) p.x = q.x;
    if (params->y_axis == 1) q.x = p.x;
	
    fragColor = _mix(_tex2DVecN(iChannel0, p.x, q.x, 15),
			        (a + b + c) / params->parameter2,
				(0.5f - (dot(color, postColor) - params->parameter5)) * amount);
	
    _tex2DVec4Write(dst, x, y, fragColor);
  }
}
]]
```

**Reply by J-i-P-i** [reply, marked as answer]
@Akascape 

I had errors with both Cuda and OpenCL.
Cuda is a little more forgiving than OpenCL, but a float2 is "required" when passing parameters to the quantize function
Cuda stumbles because only one float is available and OpenCL stumbles because a float4 is used there.
I corrected this.

Best regards

github doesn't allow files with the extension "fuse" so I renamed it to SuperGlitch.fuse.csv , please remove the csv
[SuperGlitch.fuse.csv](https://github.com/nmbr73/Shaderfuse/files/11899345/SuperGlitch.fuse.csv)

**Reply by Akascape** [reply]
@J-i-P-i Thanks a lot. Now I get it. Maybe due to different opencl version, my device was not showing those errors.

**Comment by Akascape** [comment]
@J-i-P-i I also have another question, when I add this fuse node in fusion page, a merge node also appears with it. Is there any way to stop the merge node from popping up?

**Reply by J-i-P-i** [reply]
@Akascape 

This is not due to the fuse, but is an "automatism" of Fusion. When a node is selected and a new one is added, Fusion attempts to "join" them. If one node is a source and the added node is a processing node, both are connected. If both nodes are sources, then Fusion adds a merge node. Remove activation on a node before adding.

**Comment by Akascape** [comment]
@J-i-P-i Hello again, hope you are doing well, I am currently working on this shader (https://www.shadertoy.com/view/43SGz1) and made the fuse for it. I completed all the code conversion, but I don't know why the result in fusion is not looking like the shadertoy. If you have some time to review, can you help figure it out. 
Here is the fuse file: [`LowPolyStyle.fuse`](https://github.com/Akascape/Akascape/releases/download/Test/LowPolyStyle.fuse)

Thank you :)

**Comment by J-i-P-i** [comment]
@Akascape Hello, I found three things that prevented the display for me:
1) There is no “sign()” in Cuda - so I added a define
2) The textures used (i.e. image data of the buffers) must be float32, otherwise the resolution is not sufficient for the calculation (I have attached a screenshot of how the structure should be, alternatively you can also specify the depth to float32 in the fuse)
![Screen Shot 02-20-24 at 12 02 AM](https://github.com/nmbr73/Shaderfuse/assets/78935215/6de12b9e-db3c-4c03-b260-1c320f32adb6)

3) There is a rounding problem in DCTL when processing the buffer data, so 0.5 must be added to the sampling coordinates. This is probably the most important change.

Attached is the modified fuse in a ZIP Archive.
[LowPolyStyle.zip](https://github.com/nmbr73/Shaderfuse/files/14337551/LowPolyStyle.zip)

Best regards

**Reply by Akascape** [reply]
@J-i-P-i Thank you so much for your help, I learnt a lot from you.
